pandas/_libs/tslibs/conversion.pyx

```diff
@@ -234,7 +235,9 @@ def ensure_datetime64ns(arr: ndarray, copy: bool = True):
             result = result.copy()
         return result
 
-    unit = get_datetime64_unit(arr.flat[0])
+    if arr.dtype.kind != "M":
+        raise TypeError("ensure_datetime64ns arr must have datetime64 dtype")
```

Comment:
Does this get hit internally causing another path to be executed i.e. not user facing?

Reply:
no, just being extra-careful bc if we got past here without "M" we'd segfault